docs/03-architecture/nextjs-compiler.mdx

```diff
@@ -265,14 +265,17 @@ This option has been superseded by [`optimizePackageImports`](/docs/app/api-refe
 The `define` option allows you to statically replace variables in your code at build-time.
 The option takes an object as key-value pairs, where the keys are the variables that should be replaced with the corresponding values.
 
-Use the `compiler.define` field in `next.config.js`:
+Use the `compiler.define` field in `next.config.js` to define variables for all environments (server, edge, and client). Or, use `compiler.defineServer` to define variables only for server-side (server and edge) code:
 
 ```js filename="next.config.js"
 module.exports = {
   compiler: {
     define: {
-      MY_STRING_VARIABLE: JSON.stringify('my-string'),
```

Comment:
this adds literal quotes inside the string value, which I don't think is the intention.

Reply:
The current example looks correct to me. I.e. you can `MY_VARIABLE: true` and it replaces with `true` as the value, not a string. Removing `JSON.stringify` will cause `my-string` to be inlined, which is not a valid variable and not valid code.

Reply:
`serializeDefineEnv` applies `JSON.stringify` to every value. That is then passed to https://webpack.js.org/plugins/define-plugin/#usage which has the semantics you described.

So for `nextConfig.compiler.define`, you can in fact not replace something with a global variable, but 
```js
"some_boolean": false,
"some_number": 123,
"some_str": "hello world"
```
because what `webpack.DefinePlugin` receives is
```js
"some_boolean": "false",
"some_number": "123",
"some_str": '"hello world"'
```
works as expected.